bin/akula-rpc.rs

```diff
@@ -43,15 +124,572 @@ where
             .unwrap_or(BlockNumber(0)))
     }
 
+    async fn call(&self, call_data: CallData, block_number: BlockNumber) -> RpcResult<Bytes> {
+        let block_hash = canonical_hash::read(&self.db.begin().await?, block_number)
+            .await?
+            .unwrap();
+
+        let header = header::read(&self.db.begin().await?, block_hash, block_number)
+            .await?
+            .unwrap();
+
+        let tx = &self.db.begin().await?;
+        let mut state = Buffer::new(tx, BlockNumber(0), Some(block_number));
+        let mut analysis_cache = AnalysisCache::default();
+        let block_spec = MAINNET.collect_block_spec(block_number);
+
+        let value = match call_data.value {
+            None => Default::default(),
+            Some(v) => v,
+        };
+
+        let input = match call_data.data {
+            None => Default::default(),
+            Some(i) => i,
+        };
+
+        let sender = match call_data.from {
+            None => Default::default(),
+            Some(s) => s,
+        };
+
+        let msg_with_sender = MessageWithSender {
+            message: Message::Legacy {
+                chain_id: Some(ChainId(1)),
+                nonce: 0,
+                gas_price: Default::default(),
+                gas_limit: 0,
+                action: TransactionAction::Call(call_data.to),
+                value,
+                input,
+            },
+            sender,
+        };
+
+        let gas = match call_data.gas {
```

Comment:
Call with zero gas is invalid, put a proper default instead.

Reply:
Because this line of code `res.gas_left = message.gas - gas;` will then evaluate to a positive number as if gas was gained by the execution, correct? How about setting the default to the block gas limit of the latest block, or forcing the gas parameter to be set like @battlmonstr suggested for eth_estimateGas?

Reply:
Maybe we can add a good default later? Raising an error seems reasonable if it makes your life easier for now. In any case since we strive for compatibility with geth/erigon, I'd check what's the default there. I also wonder if the RPC layer is a good place for this validation. Maybe the inner layer should validate that, and then the error bubbles back here?

Reply:
Erigon seems require gas to be set correctly for eth_call and use the block gas limit of the block identified by default_block param for eth_estimateGas when gas is missing, throwing error if neither gas or default_block is passed.